File: ios/App/View/Games/Pong/PongEngine.swift
import Foundation

/// Bot difficulty levels. `botSpeed` (px/s tracking speed) and `botJitter`
/// (how far the intercept point wobbles, in px) mirror the web `BOT_LEVELS`
/// table exactly so both platforms play identically.
enum PongDifficulty: String, CaseIterable, Identifiable {
  case easy
  case medium
  case hard
  case impossible

  var id: String { rawValue }

  var botSpeed: Double {
    switch self {
    case .easy: return 180
    case .medium: return 280
    case .hard: return 380
    case .impossible: return 620
    }
  }

  var botJitter: Double {
    switch self {
    case .easy: return 60
    case .medium: return 34
    case .hard: return 16
    case .impossible: return 0
    }
  }
}

enum PongSide: Equatable {
  case left
  case right
}

/// Discrete things that happened during one `step(dt:)` — the native twin of
/// the web game's sound/score/game-over callbacks. The renderer can use these
/// for sounds or haptics; tests use them to assert on game flow.
enum PongEvent: Equatable {
  case wallBounce
  case paddleHit(PongSide)
  case score(scorer: PongSide)
  case gameOver(winner: PongSide)
}

/// Ball state. `vx`/`vy` are a unit direction vector (as in the web version);
/// the scalar speed is derived each frame from the rally length so the ball
/// speeds up as an exchange goes on.
struct PongBall: Equatable {
  var x: Double
  var y: Double
  var vx: Double
  var vy: Double
}

/// Pure port of the web Pong simulation
/// (`web/app/src/View/Games/Pong/PongGame.tsx`). No SwiftUI here: the engine
/// is a plain state machine driven by `step(dt:)` so it can be unit-tested
/// headlessly and rendered by any frontend.
///
/// Unlike the web version there is no 2-player mode — touch devices have no
/// second keyboard player — so the right paddle is always the bot.
///
/// Randomness (serve angle, bot jitter) goes through an injected closure so
/// tests can make the simulation fully deterministic.
final class PongEngine {
  // Field geometry — must stay byte-for-byte in sync with the web constants.
  static let fieldWidth: Double = 800
  static let fieldHeight: Double = 560
  static let paddleHeight: Double = 90
  static let paddleWidth: Double = 12
  static let paddleMargin: Double = 24
  static let ballSize: Double = 12
  static let playerPaddleSpeed: Double = 420
  static let winScore = 7

  var difficulty: PongDifficulty
  /// Global speed multiplier (the web page exposes it as a slider; 1 = normal).
  var speedMultiplier: Double = 1
  /// Where the player wants the left paddle (field coordinates, from touch).
  /// `nil` means "stay put". The web version snaps the paddle to the mouse;
  /// a finger can jump across the whole field in one frame, so we chase the
  /// target at the keyboard paddle speed (420 px/s, same constant as web W/S).
  var playerTargetY: Double?

  private(set) var leftPaddleY = PongEngine.fieldHeight / 2
  private(set) var rightPaddleY = PongEngine.fieldHeight / 2
  private(set) var leftScore = 0
  private(set) var rightScore = 0
  private(set) var ball = PongBall(
    x: PongEngine.fieldWidth / 2, y: PongEngine.fieldHeight / 2, vx: 0, vy: 0
  )
  private(set) var rallyHits = 0
  private(set) var winner: PongSide?

  var isOver: Bool { winner != nil }

  /// Current scalar ball speed (px/s): base speed plus a bump per rally hit,
  /// exactly like the web `ballSpeed`.
  var currentBallSpeed: Double { 260 * speedMultiplier + Double(rallyHits) * 18 }

  private let random: () -> Double

  init(
    difficulty: PongDifficulty = .hard,
    random: @escaping () -> Double = { Double.random(in: 0..<1) }
  ) {
    self.difficulty = difficulty
    self.random = random
    newGame()
  }

  /// Full match reset (the web "New Game" / resetToken path). Pass a side to
  /// make the first serve deterministic; `nil` picks one at random like web.
  func newGame(servingToward side: PongSide? = nil) {
    leftPaddleY = Self.fieldHeight / 2
    rightPaddleY = Self.fieldHeight / 2
    leftScore = 0
    rightScore = 0
    winner = nil
    serve(toward: side ?? (random() < 0.5 ? .right : .left))
  }

  /// Advance the simulation. Mirrors the web `step` frame-for-frame: paddles
  /// move, ball moves, wall bounce, paddle bounce, scoring, re-serve.
  /// Returns the discrete events that occurred so the caller can react.
  func step(dt: Double) -> [PongEvent] {
    guard !isOver else { return [] }
    var events: [PongEvent] = []

    // Player (left): chase the touch target at the fixed paddle speed.
    if let target = playerTargetY {
      let delta = target - leftPaddleY
      leftPaddleY += sign(delta) * min(abs(delta), Self.playerPaddleSpeed * dt)
    }

    // Bot (right): only chase when the ball is coming at it; drift home
    // otherwise — same policy as web, including per-frame jitter.
    let target = ball.vx > 0
      ? ball.y + (random() - 0.5) * difficulty.botJitter
      : Self.fieldHeight / 2
    let botDelta = target - rightPaddleY
    rightPaddleY += sign(botDelta) * min(abs(botDelta), difficulty.botSpeed * dt)

    let half = Self.paddleHeight / 2
    leftPaddleY = max(half, min(Self.fieldHeight - half, leftPaddleY))
    rightPaddleY = max(half, min(Self.fieldHeight - half, rightPaddleY))

    // Ball
    let speed = currentBallSpeed
    ball.x += ball.vx * speed * dt
    ball.y += ball.vy * speed * dt

    if ball.y < Self.ballSize / 2 || ball.y > Self.fieldHeight - Self.ballSize / 2 {
      ball.vy *= -1
      ball.y = max(Self.ballSize / 2, min(Self.fieldHeight - Self.ballSize / 2, ball.y))
      events.append(.wallBounce)
    }

    let leftEdge = Self.paddleMargin + Self.paddleWidth
    let rightEdge = Self.fieldWidth - Self.paddleMargin - Self.paddleWidth
    if ball.vx < 0,
       ball.x - Self.ballSize / 2 <= leftEdge,
       ball.x > Self.paddleMargin,
       abs(ball.y - leftPaddleY) <= Self.paddleHeight / 2 + Self.ballSize / 2 {
      ball.x = leftEdge + Self.ballSize / 2
      bounceOffPaddle(paddleY: leftPaddleY)
      events.append(.paddleHit(.left))
    }
    if ball.vx > 0,
       ball.x + Self.ballSize / 2 >= rightEdge,
       ball.x < Self.fieldWidth - Self.paddleMargin,
       abs(ball.y - rightPaddleY) <= Self.paddleHeight / 2 + Self.ballSize / 2 {
      ball.x = rightEdge - Self.ballSize / 2
      bounceOffPaddle(paddleY: rightPaddleY)
      events.append(.paddleHit(.right))
    }

    // Scoring
    if ball.x < -Self.ballSize || ball.x > Self.fieldWidth + Self.ballSize {
      let scorer: PongSide = ball.x > Self.fieldWidth ? .left : .right
      if scorer == .left {
        leftScore += 1
      } else {
        rightScore += 1
      }
      events.append(.score(scorer: scorer))

      if leftScore >= Self.winScore || rightScore >= Self.winScore {
        winner = leftScore > rightScore ? .left : .right
        events.append(.gameOver(winner: winner ?? .left))
      } else {
        // Web serves toward the scorer's side after a point.
        serve(toward: scorer)
      }
    }

    return events
  }

  /// Test hook: place the ball directly to set up collision scenarios.
  func setBall(x: Double, y: Double, vx: Double, vy: Double) {
    ball = PongBall(x: x, y: y, vx: vx, vy: vy)
  }

  private func serve(toward side: PongSide) {
    // Random angle within ±45°, same distribution as web serveBall.
    let angle = (random() * 0.5 - 0.25) * Double.pi
    let direction: Double = side == .right ? 1 : -1
    ball = PongBall(
      x: Self.fieldWidth / 2,
      y: Self.fieldHeight / 2,
      vx: cos(angle) * direction,
      vy: sin(angle)
    )
    rallyHits = 0
  }

  private func bounceOffPaddle(paddleY: Double) {
    // Hit position controls the return angle, like the arcade original.
    let offset = (ball.y - paddleY) / (Self.paddleHeight / 2)
    let angle = offset * 0.75
    let direction: Double = ball.vx > 0 ? -1 : 1
    ball.vx = cos(angle) * direction
    ball.vy = sin(angle)
    rallyHits += 1
  }

  private func sign(_ value: Double) -> Double {
    if value > 0 { return 1 }
    if value < 0 { return -1 }
    return 0
  }
}

File: android/app/src/main/kotlin/com/baseapp/android/view/games/pong/PongEngine.kt
package com.baseapp.android.view.games.pong

import kotlin.math.PI
import kotlin.math.abs
import kotlin.math.cos
import kotlin.math.min
import kotlin.math.sign
import kotlin.math.sin
import kotlin.random.Random

/**
 * Pure Kotlin port of the web Pong physics
 * (`web/app/src/View/Games/Pong/PongGame.tsx`) so the exact same rules run on
 * every platform and can be unit-tested on the JVM. No Android or Compose
 * imports here — rendering and input live in `PongGameView`.
 *
 * This port is 1-player-vs-bot only: touch devices have no second keyboard
 * player, so the web game's 2P arrow-key mode is intentionally dropped.
 *
 * All coordinates are in field units (800x560, y growing downward); the view
 * scales the field to the device while preserving the aspect ratio.
 */
class PongEngine(
    var difficulty: BotLevel = BotLevel.HARD,
    /** Matches the web speed slider (0.6..1.8); 1.0 is the default. */
    var speedMultiplier: Float = 1f,
    private val random: Random = Random.Default,
) {
    /**
     * Bot tracking speed (px/s) and aim jitter per difficulty — the same
     * table as the web BOT_LEVELS so the bot feels identical on both apps.
     */
    enum class BotLevel(val speed: Float, val jitter: Float) {
        EASY(180f, 60f),
        MEDIUM(280f, 34f),
        HARD(380f, 16f),
        IMPOSSIBLE(620f, 0f),
    }

    enum class Side { LEFT, RIGHT }

    /**
     * Ball position plus a unit direction vector. Speed is applied separately
     * (see [ballSpeed]) exactly like the web game, so rally speed-ups never
     * distort the direction.
     */
    data class Ball(val x: Float, val y: Float, val vx: Float, val vy: Float)

    /** Events produced by one [step] call; null fields mean "nothing happened". */
    data class StepResult(
        val scoredBy: Side? = null,
        val gameWinner: Side? = null,
    )

    var leftPaddleY: Float = FIELD_HEIGHT / 2f
        private set
    var rightPaddleY: Float = FIELD_HEIGHT / 2f
        private set
    var leftScore: Int = 0
        private set
    var rightScore: Int = 0
        private set
    var ball: Ball = Ball(FIELD_WIDTH / 2f, FIELD_HEIGHT / 2f, 0f, 0f)
        private set
    var rallyHits: Int = 0
        private set
    var isOver: Boolean = false
        private set
    var winner: Side? = null
        private set

    private var playerTargetY: Float? = null

    /** Current ball speed in px/s: base speed plus the web's per-rally ramp. */
    val ballSpeed: Float
        get() = BASE_BALL_SPEED * speedMultiplier + rallyHits * RALLY_SPEED_STEP

    init {
        newGame()
    }

    /** Full match reset; the web equivalent is the resetToken effect. */
    fun newGame(serveDirection: Int = if (random.nextFloat() < 0.5f) 1 else -1) {
        leftPaddleY = FIELD_HEIGHT / 2f
        rightPaddleY = FIELD_HEIGHT / 2f
        leftScore = 0
        rightScore = 0
        rallyHits = 0
        isOver = false
        winner = null
        playerTargetY = null
        serve(serveDirection)
    }

    /**
     * Touch input: field-space y the player paddle should glide toward, or
     * null when there is no active touch (the paddle then holds position).
     */
    fun setPlayerTarget(y: Float?) {
        playerTargetY = y?.coerceIn(0f, FIELD_HEIGHT)
    }

    /**
     * Advance the simulation by [dtSeconds] (clamped to 50ms like the web
     * loop, so a background pause never produces a huge tunnel-through step).
     */
    fun step(dtSeconds: Float): StepResult {
        if (isOver) {
            return StepResult()
        }
        val dt = min(dtSeconds, MAX_STEP_SECONDS)

        // Player (left): glide toward the touch target, capped at the same
        // 420 px/s as the web keyboard controls — a fast drag cannot teleport
        // the paddle through the ball.
        playerTargetY?.let { target ->
            val delta = target - leftPaddleY
            leftPaddleY += sign(delta) * min(abs(delta), PLAYER_PADDLE_SPEED * dt)
        }

        // Bot (right): only chases while the ball approaches; drifts back to
        // the middle otherwise, matching the web bot.
        val botTarget = if (ball.vx > 0) {
            ball.y + (random.nextFloat() - 0.5f) * difficulty.jitter
        } else {
            FIELD_HEIGHT / 2f
        }
        val botDelta = botTarget - rightPaddleY
        rightPaddleY += sign(botDelta) * min(abs(botDelta), difficulty.speed * dt)

        val half = PADDLE_HEIGHT / 2f
        leftPaddleY = leftPaddleY.coerceIn(half, FIELD_HEIGHT - half)
        rightPaddleY = rightPaddleY.coerceIn(half, FIELD_HEIGHT - half)

        // Ball movement and wall bounce.
        val speed = ballSpeed
        var x = ball.x + ball.vx * speed * dt
        var y = ball.y + ball.vy * speed * dt
        var vx = ball.vx
        var vy = ball.vy

        if (y < BALL_SIZE / 2f || y > FIELD_HEIGHT - BALL_SIZE / 2f) {
            vy = -vy
            y = y.coerceIn(BALL_SIZE / 2f, FIELD_HEIGHT - BALL_SIZE / 2f)
        }

        // Paddle collision: hit position controls the return angle, like the
        // arcade original (offset in [-1, 1] maps to ±0.75 rad).
        val leftEdge = PADDLE_MARGIN + PADDLE_WIDTH
        val rightEdge = FIELD_WIDTH - PADDLE_MARGIN - PADDLE_WIDTH
        if (
            vx < 0 &&
            x - BALL_SIZE / 2f <= leftEdge &&
            x > PADDLE_MARGIN &&
            abs(y - leftPaddleY) <= PADDLE_HEIGHT / 2f + BALL_SIZE / 2f
        ) {
            x = leftEdge + BALL_SIZE / 2f
            val angle = deflectionAngle(y, leftPaddleY)
            vx = cos(angle)
            vy = sin(angle)
            rallyHits += 1
        }
        if (
            vx > 0 &&
            x + BALL_SIZE / 2f >= rightEdge &&
            x < FIELD_WIDTH - PADDLE_MARGIN &&
            abs(y - rightPaddleY) <= PADDLE_HEIGHT / 2f + BALL_SIZE / 2f
        ) {
            x = rightEdge - BALL_SIZE / 2f
            val angle = deflectionAngle(y, rightPaddleY)
            vx = -cos(angle)
            vy = sin(angle)
            rallyHits += 1
        }

        ball = Ball(x, y, vx, vy)

        // Scoring: the ball must fully leave the field, then the loser serves.
        if (x < -BALL_SIZE || x > FIELD_WIDTH + BALL_SIZE) {
            val leftScored = x > FIELD_WIDTH
            val scorer = if (leftScored) Side.LEFT else Side.RIGHT
            if (leftScored) {
                leftScore += 1
            } else {
                rightScore += 1
            }
            if (leftScore >= WIN_SCORE || rightScore >= WIN_SCORE) {
                isOver = true
                winner = if (leftScore > rightScore) Side.LEFT else Side.RIGHT
                return StepResult(scoredBy = scorer, gameWinner = winner)
            }
            serve(if (leftScored) -1 else 1)
            return StepResult(scoredBy = scorer)
        }

        return StepResult()
    }

    /** Serve from center toward [direction] within ±45°, like the web serveBall. */
    private fun serve(direction: Int) {
        val angle = (random.nextFloat() * 0.5f - 0.25f) * PI.toFloat()
        ball = Ball(
            x = FIELD_WIDTH / 2f,
            y = FIELD_HEIGHT / 2f,
            vx = cos(angle) * direction,
            vy = sin(angle),
        )
        rallyHits = 0
    }

    private fun deflectionAngle(ballY: Float, paddleY: Float): Float {
        val offset = (ballY - paddleY) / (PADDLE_HEIGHT / 2f)
        return offset * 0.75f
    }

    /** Test hook: pin the ball for deterministic physics assertions. */
    internal fun placeBall(x: Float, y: Float, vx: Float, vy: Float) {
        ball = Ball(x, y, vx, vy)
    }

    companion object {
        const val FIELD_WIDTH = 800f
        const val FIELD_HEIGHT = 560f
        const val PADDLE_HEIGHT = 90f
        const val PADDLE_WIDTH = 12f
        const val PADDLE_MARGIN = 24f
        const val BALL_SIZE = 12f
        const val PLAYER_PADDLE_SPEED = 420f
        const val WIN_SCORE = 7
        const val BASE_BALL_SPEED = 260f
        const val RALLY_SPEED_STEP = 18f
        const val MAX_STEP_SECONDS = 0.05f
    }
}
